pkg/sql/catalog/randgen/tables.go

```diff
@@ -0,0 +1,197 @@
+		tbName := selectNames[i]
+		id := firstID + catid.DescID(i)
+		g.genOneTable(ctx, db, sc, privs, id, tmpl, tbName)
+	}
+}
+
+// selectNamesAndTemplates selects table templates and assigns names to them.
+//
+// If the top level table pattern is "_", it computes as follows:
+// given the templates a,b,c,
+// and we are requested to produce 10 tables,
+// then select (a, "a1"), (b, "b1"), (c "c1"), (a, "a2"), (b, "b2") ... (a, "a4")
+//
+// Otherwise, non-undescore "tb", it computes as follows:
+// (a, "tb1"), (b, "tb2"), (c, "tb3"), (a, "tb4"), ... (a, "tb10")
+//
+// Additionally, the templates are shuffled so we get a different
+// subset of the templates every time (useful in particular when num
+// is smaller than the number of templates).
+func (g *testSchemaGenerator) selectNamesAndTemplates(
+	ctx context.Context, conflictNames map[string]struct{}, num int,
+) (selectTmpl []*tbTemplate, selectNames []string) {
+	// Let's consider a different template order every time.
+	g.rand.Shuffle(len(g.templates), func(i, j int) {
+		g.templates[i], g.templates[j] = g.templates[j], g.templates[i]
+	})
+
+	// How many names are we generating per template?
+	namesPerTemplate := num / len(g.templates)
+	if namesPerTemplate*len(g.templates) < num {
+		namesPerTemplate++
+	}
+
+	// Names/templates, per generator.
+	startTmpl := make([]*tbTemplate, 0, num)
+	startNames := make([][]string, 0, num)
+	if g.tbNamePat == "_" {
+		// Rotate through templates and their names.
+		for i := range g.templates {
+			tbNamePat := g.templates[i].namePat
+			cfg := g.cfg.NameGen
+			if namesPerTemplate == 1 {
+				cfg.Number = false
+				if !cfg.HasVariability() {
+					cfg.Number = true
+				}
+			}
+			ng := randident.NewNameGenerator(&cfg, g.rand, tbNamePat)
+			tbNames, err := ng.GenerateMultiple(ctx, namesPerTemplate, conflictNames)
+			if err != nil {
+				panic(genError{err})
+			}
+			startNames = append(startNames, tbNames)
+			startTmpl = append(startTmpl, &g.templates[i])
+		}
+		// Now select the name,template pairs.
+		for nameOffset := 0; len(selectNames) < num; nameOffset++ {
+			for i := range startTmpl {
+				if len(selectNames) >= num {
+					break
+				}
+				selectTmpl = append(selectTmpl, startTmpl[i])
+				selectNames = append(selectNames, startNames[i][nameOffset])
+			}
+		}
+	} else {
+		// Rotate through templates, but use a single sequence of names.
+		ng := randident.NewNameGenerator(&g.cfg.NameGen, g.rand, g.tbNamePat)
+		tbNames, err := ng.GenerateMultiple(ctx, num, conflictNames)
+		if err != nil {
+			panic(genError{err})
+		}
+		selectNames = tbNames
+		for len(selectTmpl) < num {
+			for i := range g.templates {
+				if len(selectTmpl) >= num {
+					break
+				}
+				selectTmpl = append(selectTmpl, &g.templates[i])
+			}
+		}
+	}
+	return selectTmpl, selectNames
+}
+
+func (g *testSchemaGenerator) genOneTable(
+	ctx context.Context,
+	parentDb catalog.DatabaseDescriptor,
+	parentSc catalog.SchemaDescriptor,
```

Comment:
uber nit: FYI the prevailing convention in `pkg/sql/catalog/...` is to just call these `db` and `sc`